invenio_sequencegenerator/models.py

```diff
@@ -0,0 +1,202 @@
+
+        # Check if parent sequence exists
+        names = [seq.name for seq in Sequence.query.all()]
+        parents = filter(lambda p: p in names, placeholders)
+        if len(parents) > 1:
+            raise InvalidTemplate('More than 1 parents in template')
+
+        parent = parents[0] if parents else None
+
+        db.session.add(cls(
+            name=name,
+            template=template,
+            parent=parent
+        ))
+
+    @classmethod
+    def next(cls, id, **kwargs):
+        """Get the next report number for the given sequence."""
+        with db.session.begin_nested():
+            seq = db.session.query(cls).filter_by(name=id).first() or \
+                  db.session.query(cls).filter_by(template=id).first()
+            if seq is None:
+                raise NoSuchSequence()
+
+            try:
+                template = seq.template.format(**kwargs)
+            except KeyError:
+                raise InvalidCall('Improper arguments for string formatting')
+
+            if seq.parent:
+                # Check if parent's report number exists in PID store
+                try:
+                    get_distribution('invenio-pidstore')
+                    # from invenio_pidstore.errors import PIDDoesNotExistError
+                    from invenio_pidstore.models import PersistentIdentifier
+
+                    try:
+                        PersistentIdentifier.query.filter_by(
+                            pid_value=template
+                        ).one()
+                    except NoResultFound:
+                        raise InvalidCall('Parent PID does not exist')
+                except DistributionNotFound:
+                    pass
+
+                return ReportNumber.next_or_create(template)
+            else:
+                return ReportNumber.next_or_create(template)
+
+
+class ReportNumber(db.Model):
+    """Stores generated report numbers."""
+
+    __tablename__ = 'sequencegenerator_reportnumber'
+
+    template = db.Column(db.String, nullable=False,
+                         primary_key=True, index=True)
+
+    counter = db.Column(db.Integer, nullable=False)
+
+    @classmethod
+    def create(cls, template, counter=1):
+        """Create a new report number."""
+        with db.session.begin_nested():
+            rec = cls(
+                template=template,
+                counter=counter
+            )
+            db.session.add(rec)
+        return rec
+
+    @classmethod
+    def next_or_create(cls, template):
+        """Get the next report number for the given template."""
+        with db.session.begin_nested():
+            rec = db.session.query(cls).filter_by(template=template).first()
+
+            if rec is None:
+                rec = ReportNumber.create(template)
+            else:
+                rec.counter += 1
+
+        return rec.template.format(counter=rec.counter)
+
+
+#
+# User-level API
+#
+def new_sequence(name, template):
```

Comment:
IMHO there is no need in defining functional API if we already have classmethods.

**Context:**
We've removed e.g. `get_record` function in favor of `Record.get_record`.